visitors/http_rest_diff/http_rest_spec_diff_visitor.go

```diff
@@ -0,0 +1,550 @@
+
+	EnterAddedOrRemovedList(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.List) Cont
+	LeaveAddedOrRemovedList(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.List, cont Cont) Cont
+
+	EnterAddedOrRemovedOneOf(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.OneOf) Cont
+	LeaveAddedOrRemovedOneOf(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.OneOf, cont Cont) Cont
+
+	EnterAddedOrRemovedOptional(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Optional) Cont
+	LeaveAddedOrRemovedOptional(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Optional, cont Cont) Cont
+
+	EnterAddedOrRemovedPrimitive(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Primitive) Cont
+	EnterChangedPrimitive(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Primitive) Cont
+	LeaveAddedOrRemovedPrimitive(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Primitive, cont Cont) Cont
+	LeaveChangedPrimitive(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Primitive, cont Cont) Cont
+
+	EnterAddedOrRemovedStruct(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Struct) Cont
+	LeaveAddedOrRemovedStruct(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right *pb.Struct, cont Cont) Cont
+}
+
+// An HttpRestSpecDiffVisitor with convenience functions for entering and
+// leaving nodes with diffs.
+type DefaultHttpRestSpecDiffVisitor interface {
+	HttpRestSpecDiffVisitor
+
+	EnterDiff(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right interface{}) Cont
+	LeaveDiff(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right interface{}, cont Cont) Cont
+
+	// Delegates to EnterDiff by default.
+	EnterAddedOrRemovedNode(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right interface{}) Cont
+
+	// Delegates to EnterDiff by default.
+	EnterChangedNode(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right interface{}) Cont
+
+	// Delegates to LeaveDiff by default.
+	LeaveAddedOrRemovedNode(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right interface{}, cont Cont) Cont
+
+	// Delegates to LeaveDiff by default.
+	LeaveChangedNode(self interface{}, ctx http_rest.HttpRestSpecPairVisitorContext, left, right interface{}, cont Cont) Cont
+}
+
+// An HttpRestSpecDiffVisitor implementation. This does not traverse into the
+// children of nodes that were added, removed, or changed.
+type DefaultHttpRestSpecDiffVisitorImpl struct {
+	http_rest.DefaultHttpRestSpecPairVisitor
+}
+
+var _ DefaultHttpRestSpecDiffVisitor = &DefaultHttpRestSpecDiffVisitorImpl{}
```

Comment:
A variant that also works is:

```
        var _ DefaultHttpRestSpecDiffVisitor = (DefaultHttpRestSpecDiffVisitorImpl *)(nil)
```

It looks like Vault uses both varieties.

Reply:
In 8147936.